code/controllers/subsystem/tts.dm

```diff
@@ -0,0 +1,223 @@
+		var/timeout_time = current_message[START_TIME_INDEX] + message_timeout
+		var/identifier = current_message[IDENTIFIER_INDEX]
+		cached_voices -= identifier
+		if(response.errored || timeout_time < world.time)
+			continue
+
+		var/sound/new_sound = new("tmp/tts/[identifier].ogg")
+		for(var/target in current_message[TARGET_INDEX])
+			play_tts(target["target"], new_sound, target["language"], target["local"], target["range"])
+		if(MC_TICK_CHECK)
+			return
+
+/proc/cmp_word_length_asc(list/a, list/b)
+	return length(b[MESSAGE_INDEX]) - length(a[MESSAGE_INDEX])
+
+#define ADD_TARGET_TO_STRUCT(tts_struct, target, language, local, range) ##tts_struct[TARGET_INDEX] += list(list("target" = ##target, "language" = ##language, "local" = ##local, "range" = ##range))
+
+/datum/controller/subsystem/tts/proc/queue_tts_message(target, message, datum/language/language, speaker, filter, local = FALSE, message_range = 7)
+	if(!tts_enabled)
+		return
+
+	var/static/regex/contains_alphanumeric = regex("\[a-zA-Z0-9]")
+	// If there is no alphanumeric char, the output will usually be static, so
+	// don't bother sending
+	if(contains_alphanumeric.Find(message) == 0)
+		return
+
+	var/shell_scrubbed_input = tts_speech_filter(message)
+	shell_scrubbed_input = copytext(shell_scrubbed_input, 1, 300)
+	var/identifier = sha1(speaker + filter + shell_scrubbed_input)
+	var/cached_voice = cached_voices[identifier]
+	if(islist(cached_voice))
+		ADD_TARGET_TO_STRUCT(cached_voice, target, language, local, message_range)
+		return
+	else if(fexists("tmp/tts/[identifier].ogg"))
+		var/sound/new_sound = new("tmp/tts/[identifier].ogg")
+		play_tts(target, new_sound, language, local, message_range)
+		return
+	if(!(speaker in available_speakers))
+		return
+
+	var/list/headers = list()
+	headers["Content-Type"] = "application/json"
+	headers["Authorization"] = CONFIG_GET(string/tts_http_token)
+	var/datum/http_request/request = new()
+	var/file_name = "tmp/tts/[identifier].ogg"
+	request.prepare(RUSTG_HTTP_METHOD_GET, "[CONFIG_GET(string/tts_http_url)]/tts?voice=[speaker]&identifier=[identifier]&filter=[url_encode(filter)]", json_encode(list("text" = shell_scrubbed_input)), headers, file_name)
+	// This'll probably be better off datumized in the future, but it's not necessary to do right now
+	var/list/data = list(
+		// TARGET_INDEX = 1
+		list(),
+		// IDENTIFIER_INDEX = 2
+		identifier,
+		// START_TIME_INDEX = 3
+		world.time,
+		// REQUEST_INDEX = 4
+		request,
+		// MESSAGE_INDEX = 5
+		shell_scrubbed_input,
+	)
+	ADD_TARGET_TO_STRUCT(data, target, language, local, message_range)
+	cached_voices[identifier] = data
+	if(length(in_process_tts_messages) < max_concurrent_requests)
+		request.begin_async()
+		in_process_tts_messages += list(data)
+		sortTim(in_process_tts_messages, GLOBAL_PROC_REF(cmp_word_length_asc))
+	else
+		queued_tts_messages += list(data)
+		sortTim(queued_tts_messages, GLOBAL_PROC_REF(cmp_word_length_asc))
```

Comment:
did you decide against /datum/heap for this btw? or was that missed

Reply:
didn't know it existed, thanks